core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,601 +220,1206 @@ private SqlQueryScheduler(
-        for (SubPlan planNode : Traverser.forTree(SubPlan::getChildren).breadthFirst(planTree)) {
-            PlanFragment fragment = planNode.getFragment();
-            SqlStageExecution stageExecution = createSqlStageExecution(
-                    getStageId(session.getQueryId(), fragment.getId()),
-                    fragment,
-                    extractTableInfo(session, metadata, fragment),
-                    taskFactory,
-                    session,
-                    summarizeTaskInfo,
-                    nodeTaskMap,
-                    executor,
-                    schedulerStats);
-            result.put(fragment.getId(), stageExecution);
-        }
-        return result.build();
-    }
+        coordinatorStagesScheduler.schedule();
 
-    private static Map<PlanNodeId, TableInfo> extractTableInfo(Session session, Metadata metadata, PlanFragment fragment)
-    {
-        return searchFrom(fragment.getRoot())
-                .where(TableScanNode.class::isInstance)
-                .findAll()
-                .stream()
-                .map(TableScanNode.class::cast)
-                .collect(toImmutableMap(PlanNode::getId, node -> getTableInfo(session, metadata, node)));
+        Optional<DistributedStagesScheduler> distributedStagesScheduler = createDistributedStagesScheduler(currentAttempt.get());
+        distributedStagesScheduler.ifPresent(scheduler -> distributedStagesSchedulingTask = executor.submit(scheduler::schedule, null));
     }
 
-    private static TableInfo getTableInfo(Session session, Metadata metadata, TableScanNode node)
+    private synchronized Optional<DistributedStagesScheduler> createDistributedStagesScheduler(int attempt)
     {
-        TableSchema tableSchema = metadata.getTableSchema(session, node.getTable());
-        TableProperties tableProperties = metadata.getTableProperties(session, node.getTable());
-        return new TableInfo(tableSchema.getQualifiedName(), tableProperties.getPredicate());
-    }
+        if (queryStateMachine.isDone()) {
+            return Optional.empty();
+        }
+        DistributedStagesScheduler distributedStagesScheduler = PipelinedDistributedStagesScheduler.create(
+                queryStateMachine,
+                schedulerStats,
+                nodeScheduler,
+                nodePartitioningManager,
+                stages,
+                coordinatorStagesScheduler,
+                executionPolicy,
+                failureDetector,
+                schedulerExecutor,
+                splitSourceFactory,
+                splitBatchSize,
+                dynamicFilterService,
+                tableExecuteContextManager,
+                retryPolicy,
+                attempt);
+        this.distributedStagesScheduler.set(distributedStagesScheduler);
+        distributedStagesScheduler.addStateChangeListener(state -> {
+            if (queryStateMachine.getQueryState() == QueryState.STARTING && state != DistributedStagesSchedulerState.PLANNED) {
```

Comment:
nit: `state != DistributedStagesSchedulerState.PLANNED` -> `state.ordinal() >= DistributedStagesSchedulerState.RUNNING.ordinal()`

Reply:
Or add `state.isRunningOrFinished()` method and use it.

Reply:
Added `isRunningOrDone`